ppcs/ppcTODO-signature-named-parameters.md

```diff
@@ -0,0 +1,153 @@
+
+## Security Implications
+
+There are no anticipated security concerns with expanding the way that subroutines process parameters in this fashion.
+
+## Examples
+
+As the intention of this syntax addition is to make existing code practice more consise and simple to write, it would be illustrative to compare pairs of functions written in the newly proposed vs. the existing style.
+
+One example comes from [`IPC::MicroSocket::Server`](https://metacpan.org/pod/IPC::MicroSocket::Server). This is currently using the `Sublike::Extended` module (see "Prototype Implementation" below) to provide the syntax. The actual module also uses `Object::Pad` to provide the `method` keyword; this example is paraphrased to avoid that. This module requires a `path` named argument, and optionally takes a `listen` argument, defaulting its value to 5 if the caller did not provide a defined value.
+
+```perl
+extended sub new_unix ( $class, :$path, :$listen //= 5 )
+{
+    ...
+}
+```
+
+This replaces the previous version of the code which handled arguments by the more traditional approach of assigning into a hash:
+
+```perl
+sub new_unix ( $class, %args )
+{
+    my $path   = $args{path};
+    my $listen = $args{listen} // 5;
+    ...
+}
+```
+
+Already the new code is shorter and more consise. Additionally it contains error checking that complains about missing mandatory keys, or unrecognised keys, which the previous version of the code did not include.
+
+Another example, this time from [`Text::Treesitter::QueryCursor`](https://metacpan.org/pod/Text::Treesitter::QueryCursor). This method takes an optional argument named `multi`, tested for truth. If absent it should default to false.
+
+```perl
+sub next_match_captures ( $self, :$multi = 0 )
+{
+    ...
+}
+```
+
+The previous version of this code did include complaints about unrecognised keys, and was rather longer because of it:
+
+```perl
+sub next_match_captures ( $self, %options )
+{
+    my $multi = delete $options{multi};
+    keys %options and
+        croak "Unrecognised options to ->next_captures: " . join( ", ", keys %options );
+
+    ...
+}
+```
+
+## Prototype Implementation
+
+The [`XS::Parse::Sublike`](https://metacpan.org/pod/XS::Parse::Sublike) module contains parsing to allow third-party syntax modules to parse subroutine-like constructions, and includes a parser for named parameters already having this syntax. These are also made available to regular subroutines via the `extended` keyword provided by [`Sublike::Extended`](https://metacpan.org/pod/Sublike::Extended).
+
+## Future Scope
+
+* If the Metaprogramming API (PPC0022) gains introspection abilities to enquire about subroutine signature parameters, further consideration will need to be made in that API on how to represent the extra kinds of parameters added by this specification.
+
```

Comment:
Is there potentially scope for using this outside of function params to de-structure paired lists in other contexts?

```perl
my @users = ( 
    { user => 'bob',    pass => 'abc123' },
    { user => 'alice',  pass => 'abc123' },
    { user => 'ted',    pass => 'abc123' },
    { user => 'caroll', pass => 'abc123' },
);

for my $user_entry (@users) { 
   my (:$user, :$pass) = $_->%*;
   say "$user password is $pass";
}
```
Or
```perl
use List::Util qw(mesh);

sub map_users_to_accounts(@users) { 
   my @accounts = get_accounts_for(@users);
   return mesh \@users, \@accounts;
}

my (:$bob, :$alice, :$ted, :$caroll) = map_users_to_accounts(@users);
```
And can we use it in the opposite direction?
```perl
my $path = get_path();
new_unix( $class, :$path )
```
and 
```perl
my ($bob, $alice, $ted, $caroll) = get_users();
my %accounts = (:$bob, :$alice, :$ted, :$caroll);
```

Reply:
i second this idea, tho i think it's out of scope for this PPC persay; it's good to note that the design here is forward compatible with this type of syntax so maybe we should leave a note about that

Reply:
Certainly there is, as a separate idea. In fact I've been working on a `Syntax::Operator::Assign`, not yet on CPAN, to do exactly that kind of thing. Its LHS is supposed to accept anything that a signature would allow; in effect it means that function invocation is just an assignment from the caller's arguments into the signature list. Having named parameters in signatures is definitely a stepping-stone towards that, among other ideas.

Reply:
This is also out of scope and for more general use, but what about the use of the experimental `refaliasing`/`declared_refs` feature when it comes to use with parameter syntax? I know that @leonerd has <https://metacpan.org/pod/Signature::Attribute::Alias> which does aliasing via a parameter attribute syntax.